### ISSUE TITLE
[cpp] Extending IMap interface doesn't work as I would expect

### ISSUE THREAD

**Issue by azrafe7** [azrafe7/hxOrderedMap#2]
**UPDATE(02 nov 2018):** I have a work-around in place that makes the lib work on cpp too. Still not totally happy with it and trying to reduce the case so I can submit a proper issue upstream.

**ORIGINAL POST(31 oct 2018):**
This most probably is my bad somewhere, although I haven't found a way to solve it yet.
Other targets are ok with it.

Currently I'm just including the `IMap` interface into the `IOrderedMap` one (for **cpp**).

**Comment by azrafe7** [comment]
Error during compilation:

```
Error: TestOrderedStringMap.cpp
include\OrderedStringMapImpl.h(60): error C2535: 'void OrderedStringMapImpl_obj::set_41d0cb46(Dynamic,Dynamic)': member function already defined or declared
include\OrderedStringMapImpl.h(50): note: see declaration of 'OrderedStringMapImpl_obj::set_41d0cb46'
```

It seems that hxcpp generates 2 identical definitions for the set() method. Here's lines 48-63 from OrderedStringMapImpl.h:


```
48    bool _hx_isInstanceOf(int inClassId);
49    void *_hx_getInterface(int inHash);
50    void set_41d0cb46( ::Dynamic k, ::Dynamic v);
51 
52    ::cpp::VirtualArray keysCopy_615cc599();
53 
54    ::Dynamic copy_73f727cb();
55 
56    bool remove_5889326f( ::Dynamic k);
57 
58    bool exists_5889326f( ::Dynamic k);
59 
60    void set_41d0cb46( ::Dynamic k, ::Dynamic v);
61 
62     ::Dynamic get_dc94b8fa( ::Dynamic k);
63    ::String __ToString() const { return HX_("OrderedStringMapImpl",3e,92,55,81); }
```

Travis log: https://travis-ci.org/azrafe7/hxOrderedMap/builds/449505376

**Comment by azrafe7** [comment]
hl seems to fail too (BUT at run-time, rather than compile-time) with:

```
Uncaught exception: Can't cast #OrderedObjectMapImpl to #OrderedStringMapImpl
Called from $TestAll.main(TestAll.hx:101)
Called from fun$796(?:1)
```

**EDIT:** this was my fault, definitely not a bug on **hl** (can still reproduce the issue on **cpp** though)!
With hashlink it failed due to an oversight related to a commit before c1e555fd10371a8547526548ecc1f1526d8303f1 (a commit that I didn't test with hl, but was giving errors for java and neko and seemed unrelated).

**Comment by azrafe7** [comment]
Now I'm seeing some problems with cs (travis log: https://travis-ci.org/azrafe7/hxOrderedMap/builds/449555603):

```
src/OrderedMap.cs(27,16): error CS0121: The call is ambiguous between the following methods or properties: `IOrderedMap.keys()' and `haxe.IMap.keys()'
src/IOrderedMap.cs(33,9): (Location of the symbol related to previous error)
src/haxe/Constraints.cs(30,10): (Location of the symbol related to previous error)
Compilation failed: 1 error(s), 0 warnings
Compilation error
Native compilation failed
Error: Build failed
```

**EDIT:** fixed in a _somewhat hacky way_ in 7ae4396a71c871f9e3bb029958ebfa06515ba69b

Leaving the issue open for now.
